### PR TITLE
tests: prepare for ssh task comms

### PR DESCRIPTION
**Pull request by oliver-sanders**
* use extglobs for platform pattern matching comms:?(tcp|ssh)
* upgrade cylc message tests to run on remote platforms (as well as
  local ones)
* set cylc message tests to run for ssh task comms (as well as tcp)
* set fileinstall tests to run for ssh (as well as tcp)
* improve test battery documentation

This should make testing #3327 easier, deals with some of the less well documented aspects of the functional testing framework.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.
- [x] No dependency changes.

### CODE REVIEW

File: tests/functional/README.md [hunk 1]
```diff
@@ -32,31 +32,100 @@ command line interfaces / outputs, etc.
...
+
+By default tests require the platform `_local_background_indep_tcp`.
+
+If you want your test to run on any other platform export an environment
+variable called `REQUIRE_PLATFORM` *before* the `test_header` is sourced.
+
+This variable should define a test's requirements of a platform e.g:
+
+```bash
+# require a remote platform with a shared filesystem ans tcp task comms
```
Comment (MetRonnie) [inline tests/functional/README.md:104]:
```suggestion
# require a remote platform with a shared filesystem and tcp task comms
```

File: tests/functional/README.md [hunk 2]
```diff
@@ -32,31 +32,100 @@ command line interfaces / outputs, etc.
...
+```
+
+Bash extglob pattern matching can be provided
+(https://www.gnu.org/software/bash/manual/html_node/Pattern-Matching.html)
+e.g:
+
+```bash
+# run for all job runners with tcp comms on remote platforms
+export REQUIRE_PLATFORM='loc:remote runner:* comms:tcp'
+# run for tcp and ssh comms on remote platformexprot REQUIRE_PLATFORM='loc:remote comms:?(tcp|ssh)'
```
Comment (MetRonnie) [inline tests/functional/README.md:115]:
```suggestion
# run for tcp and ssh comms on remote platforms
```

File: etc/conf/ssh_config [hunk 3]
```diff
@@ -13,6 +13,9 @@ Host _remote*
 Host _remote_background_indep_tcp
     Port 42220
 
+Host _remote_background_indep_ssh
+    Port 42220
+
```
Comment (MetRonnie) [inline etc/conf/ssh_config:18]: Is this supposed to be different to `_remote_background_indep_tcp`?

Comment (datamel) [reply etc/conf/ssh_config:18]: Same port - Oliver told me on Friday :)

Comment (oliver-sanders) [reply etc/conf/ssh_config:18]: Yep, SSH defaults to use the TCP protocol anyway so we can use the same container for both.

> **Note:** We use a different container for the polling tests which should not be capable of TCP comms just to make sure. It's a bit over the top but should give us good confidence that polling works correctly.

I've added `_remote_background_indep_ssh` as a separate host to fit in with the same approach (easier to handle in actions, etc) but we could just use the same host here.

We could add another container (duplicate of the TCP one) to the swarm to fully separate them.

**Review by MetRonnie — APPROVED** [review]

**Review by datamel — APPROVED** [review]
Test failures unrelated, changes look good to me. Thanks Oliver!

**Review by hjoliver — APPROVED** [review]
Tests passing now :+1:

**Pull request merged**